annlite/core/codec/base.py

```diff
@@ -36,3 +39,36 @@ def is_trained(self):
 
     def _check_trained(self):
         assert self.is_trained is True, f'{self.__class__.__name__} requires training'
+
+
+class BaseTrainedPQ(ABC):
```

Comment:
I think right now there is no need to have another abstract class if there will only be one implementation

Reply:
My intention here was to make a loose coupling between the PQ and HNSW backend. Since the backend only needs a PQ that can encode(quantize) the embedding, and `BaseCodec` requires too much. The added `BaseTrainedPQ ` offers minimum interfaces for the PQ used in the HNSW backend.
The vanilla PQ implementation in `annlite.core.codec.pq` can easily support `BaseTrainedPQ`, so I added related methods. However, we hope any well-trained PQ algorithm(such as OPQ) can be used in the HNSW backend with the simplest wrapper.
(but it's over-designed here if we assume users don't want to customize their own PQ for the HNSW backend)

Reply:
I feel this is early overdesign yes